psyneulink/library/compositions/pytorchcomponents.py

```diff
@@ -0,0 +1,257 @@
+        
+        for i in range(len(mech_input.type.pointee.elements[0])):
+            pnlvm.helpers.printf_float_array(builder, builder.gep(mech_input, [ctx.int32_ty(0), ctx.int32_ty(0), ctx.int32_ty(i)]), prefix=f"{self} input:\n", override_debug=False)
+        pnlvm.helpers.printf_float_array(builder, builder.gep(mech_output, [ctx.int32_ty(0), ctx.int32_ty(0)]), prefix=f"{self} output:\n", override_debug=False)
+
+        return mech_output
+
+    def execute_derivative_func_llvm(self, ctx, builder, mech_input):
+        derivative_func = ctx.import_llvm_function(
+                    bin_function_derivative_creator(ctx, self._mechanism, context=self._context).name)
+        return gen_inject_unary_function_call(ctx, builder, derivative_func, mech_input)
+
+    def get_afferent_idx(self, afferent_node):
+        return [proj.receiver for proj in self.afferents].index(afferent_node)
+
+    def get_efferent_idx(self, efferent_node):
+        return [proj.sender for proj in self.efferents].index(efferent_node)
+
+    def __repr__(self): 
+        return "PytorchWrapper for: "+self._mechanism.__repr__() 
+
+class PytorchProjection():
+    """
+    An interpretation of a projection as an equivalent pytorch object
+    """
+    def __init__(self, projection, component_idx, port_idx, device, sender=None, receiver=None, context=None):
+        self._projection = projection
+        self._idx = component_idx
+        self._context = context
+
+        self.sender = sender
+        self.receiver = receiver
+        self._port_idx = port_idx
+
+        matrix = projection.parameters.matrix.get(
+                            context=context)
+        if matrix is None:
+            matrix = projection.parameters.matrix.get(
+                context=None
+            )
+        self.matrix = torch.nn.Parameter(torch.tensor(matrix.copy(),
+                                   device=device,
+                                   dtype=torch.double))
+
+        if projection.learnable is False:
+            self.matrix.requires_grad = False
+
+    def execute(self, variable):
+        # print(f"{self.sender._mechanism} -> {self.receiver._mechanism} input:\n{variable}")
+        # print(f"{self.sender._mechanism} -> {self.receiver._mechanism} mat:\n{self.matrix}")
+        # print(f"{self.sender._mechanism} -> {self.receiver._mechanism} output:\n{torch.matmul(variable, self.matrix)}")
+
+        return torch.matmul(variable, self.matrix)
+
+    def _extract_llvm_matrix(self, ctx, builder, params):
+        proj_params = builder.gep(params, [ctx.int32_ty(0),
+                                           ctx.int32_ty(1),
+                                           ctx.int32_ty(self._idx)])
+
+        dim_x, dim_y = self.matrix.detach().numpy().shape
+        proj_matrix = pnlvm.helpers.get_param_ptr(builder, self._projection, proj_params, "matrix")
+        proj_matrix = builder.bitcast(proj_matrix, pnlvm.ir.types.ArrayType(
```

Comment:
pls no `bitcast`. It breaks and invalidates a lot of things and should not be used anywhere.

Reply:
this would require removing the dimensionality assertions in the `_gen_inject_<op>` helpers, as previously used, since matrix is a flattened parameter (and hence is of type double*, instead of an array of array of doubles)

not sure if that would be preferable

Reply:
will flag as future cleanup